src/providers/google/calendar-event.ts

```diff
@@ -0,0 +1,390 @@
+
+  public getCalendarClient(): calendar_v3.Calendar {
+    const oAuth2Client = this.getGoogleAuth();
+    return google.calendar({ version: "v3", auth: oAuth2Client });
+  }
+
+  protected async buildCalendarList(): Promise<SchemaCalendar[]> {
+    const calendarClient = this.getCalendarClient();
+
+    let calendarList: SchemaCalendar[] = [];
+
+    let nextPageToken: string | undefined;
+    let query: calendar_v3.Params$Resource$Calendarlist$List = {
+      maxResults: MAX_BATCH_SIZE,  // Fetch in batches up to the max limit
+      pageToken: nextPageToken,
+    };
+
+    // Loop through paginated results
+    do {
+      const response = await calendarClient.calendarList.list(query);
+
+      for (const calendar of response.data.items || []) {
+        // Extract essential details for the calendar entry
+        const calendarId = calendar.id;
+
+        if (!calendarId) {
+          this.emit("log", {
+            level: SyncProviderLogLevel.DEBUG,
+            message: `Invalid calendar ID. Ignoring this calendar.`,
+          });
+          continue;
+        }
+
+        const summary = calendar.summary ?? "No calendar title";
+        let timeZone = calendar.timeZone;
+
+        if (!timeZone) {
+          this.emit("log", {
+            level: SyncProviderLogLevel.DEBUG,
+            message: `Invalid timezone for calendar ${calendarId}. Ignoring this calendar.`,
+          });
+          continue;
+        }
+
+        timeZone = CalendarHelpers.getUTCOffsetTimezone(timeZone);
+        const description = calendar.description;
+        const location = calendar.location;
+        const insertedAt = new Date().toISOString();
+
+        const group: SchemaCalendar = {
+          _id: this.buildItemId(calendarId),
+          name: summary,
+          sourceAccountId: this.provider.getAccountId(),
+          sourceApplication: this.getProviderApplicationUrl(),
+          sourceId: calendarId,
+          timezone: timeZone,
+          description,
+          location,
+          insertedAt,
+          sourceData: calendar,
+          schema: CONFIG.verida.schemas.CALENDAR,
+        };
+
+        calendarList.push(group);
+      }
+
+      nextPageToken = response.data.nextPageToken; // Update the pageToken to fetch the next batch
+      query.pageToken = nextPageToken;
+    } while (nextPageToken);
+
+
+    return calendarList;
+  }
+
+  protected async fetchEventRange(
+    calendar: SchemaCalendar,
+    range: ItemsRange,
+    apiClient: calendar_v3.Calendar
+  ): Promise<SchemaEvent[]> {
+    const events: SchemaEvent[] = [];
+
+    let query: calendar_v3.Params$Resource$Events$List = {
+      calendarId: calendar.sourceId,
+      maxResults: this.config.eventsPerCalendarLimit,
+      singleEvents: true,
+      orderBy: "startTime"
+    };
+
+    if (range.startId && !isNaN(Date.parse(range.startId))) {
+      query.timeMin = new Date(range.startId).toISOString();
+    }
+
+    if (range.endId && !isNaN(Date.parse(range.endId))) {
```

Comment:
`range.endId` should be the ID of the record that is at the end of the range and if it is hit, then the loop ends. No need to limit in the Google API query.

Reply:
Fixed